### PR TITLE
feat: add pure logic for getting smart-contracts addresses, hashes and policy ids

### PR DESCRIPTION
**Pull request by LGLO**
# Description

Code to get what `pc-contracts-cli addresses` return. It is a mix of UPLC and CSL.
It awaits v6.2.2 release of smart-contracts.
Also at v7.0.0 more scripts will use versioning system, so the logic will change towards what native-token uses at the moment.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages.
- [x] New tests are added if needed and existing tests are updated.
- [x] Relevant logging and metrics added
- [x] CI passes. See note on CI.
- [ ] Any changes are noted in the `changelog.md` for affected crate
- [x] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

File: toolkit/offchain/src/untyped_plutus.rs [hunk 1]
```diff
@@ -10,31 +9,56 @@ use uplc::plutus_data;
 /// Currently there is no other known option to apply parameters to plutus script in Rust.
 ///
 /// Parameters:
-/// * `params` - parameters to apply to the script
+/// * `param` - single PlutusData parameter to apply to the script
 /// * `plutus_script_raw` - raw plutus script in CBOR format, like in `RawScripts.purs` in smart-contracts repository
-pub(crate) fn apply_params_to_script<T: ToDatum>(
-	params: &T,
+pub fn apply_params_to_script<T: ToDatum>(
```
Comment (kpinter-iohk) [inline toolkit/offchain/src/untyped_plutus.rs:14]:
```suggestion
pub fn apply_param_to_script<T: ToDatum>(
```

File: toolkit/offchain/src/scripts_data.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+				)),
+			},
+		}
+	}
+
+	// Expected values are the ones obtained from pc-contracts-cli for the TEST_PARAMS.
+	#[test]
+	fn test_get_scripts_data0() {
+		let actual =
+			crate::scripts_data::get_scripts_data0(TEST_PARAMS, NetworkIdKind::Testnet).unwrap();
```
Comment (kpinter-iohk) [inline toolkit/offchain/src/scripts_data.rs:292]:
```suggestion
			crate::scripts_data::get_scripts_data(TEST_PARAMS, NetworkIdKind::Testnet).unwrap();
```

**Review by kpinter-iohk — APPROVED** [review]
Minor comment, otherwise looks good

File: toolkit/offchain/src/untyped_plutus.rs [hunk 3]
```diff
@@ -10,31 +9,56 @@ use uplc::plutus_data;
...
-		plutus_data(&minicbor::to_vec(params.to_datum()).expect("to_vec is Infallible"))
-			.expect("trasformation from PC Datum to pallas PlutusData can't fail");
+	let param = datum_to_uplc_plutus_data(param);
+	apply_uplc_params_to_script(&[&param], plutus_script_raw)
+}
 
+/// Parameters:
+/// * `params` - list of PlutusData parameters to apply to the script, not necessarily the "partner-chain params"
+/// * `plutus_script_raw` - raw plutus script in CBOR format, like in `RawScripts.purs` in smart-contracts repository
+pub fn apply_uplc_params_to_script(
```
Comment (kpinter-iohk) [inline toolkit/offchain/src/untyped_plutus.rs:25]:
```suggestion
pub fn apply_params_to_script(
```

**Review by AmbientTea — APPROVED** [review]

File: toolkit/offchain/src/untyped_plutus.rs (hunk 1 above)
Comment (LGLO) [reply toolkit/offchain/src/untyped_plutus.rs:14]: Good catch!

File: toolkit/offchain/src/untyped_plutus.rs (hunk 3 above)
Comment (LGLO) [reply toolkit/offchain/src/untyped_plutus.rs:25]: Yeah, I'll go for it. `uplc` was used to make distinction from `uplc::PlutusData` and our `ToDatum` At some point we should settle with only one implementation.

**Pull request merged**